libdevice/sanitizer_utils.cpp

```diff
@@ -442,6 +442,57 @@ void __asan_report_access_error(uptr addr, uint32_t as, size_t size,
                               memory_type, error_type, is_recover);
 }
 
+void __asan_report_misalign_error(uptr addr, uint32_t as, size_t size,
+                                  bool is_write, uptr poisoned_addr,
+                                  const char __SYCL_CONSTANT__ *file,
+                                  uint32_t line,
+                                  const char __SYCL_CONSTANT__ *func,
+                                  bool is_recover = false) {
+
+  auto *shadow = (__SYCL_GLOBAL__ s8 *)MemToShadow(addr, as);
+  while (*shadow >= 0) {
+    ++shadow;
+  }
+  int shadow_value = *shadow;
+
+  DeviceSanitizerErrorType error_type = DeviceSanitizerErrorType::MISALIGNED;
+  DeviceSanitizerMemoryType memory_type;
+
+  switch (shadow_value) {
+  case kUsmDeviceRedzoneMagic:
+  case kUsmDeviceDeallocatedMagic:
+    memory_type = DeviceSanitizerMemoryType::USM_DEVICE;
+    break;
+  case kUsmHostRedzoneMagic:
+  case kUsmHostDeallocatedMagic:
+    memory_type = DeviceSanitizerMemoryType::USM_HOST;
+    break;
+  case kUsmSharedRedzoneMagic:
+  case kUsmSharedDeallocatedMagic:
+    memory_type = DeviceSanitizerMemoryType::USM_SHARED;
+    break;
+  case kPrivateLeftRedzoneMagic:
+  case kPrivateMidRedzoneMagic:
+  case kPrivateRightRedzoneMagic:
+    memory_type = DeviceSanitizerMemoryType::PRIVATE;
+    break;
+  case kMemBufferRedzoneMagic:
+    memory_type = DeviceSanitizerMemoryType::MEM_BUFFER;
+    break;
+  case kSharedLocalRedzoneMagic:
+    memory_type = DeviceSanitizerMemoryType::LOCAL;
+    break;
+  case kDeviceGlobalRedzoneMagic:
+    memory_type = DeviceSanitizerMemoryType::DEVICE_GLOBAL;
+    break;
+  default:
+    memory_type = DeviceSanitizerMemoryType::UNKNOWN;
+  }
```

Comment:
move this part of code into a helper function so that it can be reused in __asan_report_access_error as well?

Reply:
done